### PR TITLE
Clustering: Interface for Peers in other packages

### PR DESCRIPTION
**Pull request by gotjosh**
A Peer as defined by the `cluster` package represents the node in the
cluster. It is used in other packages to know the status of all of the
members or how long should we wait to know if a notification has already fired.

In Cortex, we'd like to implement a slightly different way of
clustering (using gRPC for communication and a
hash ring for node discovery).

This is a small change to support that by changing the consumer of other
packages to an interface.

~~Silences and Notification channels don't need an interface as they take
a `func([]byte) error` as a parameter.~~

Edit: Ended up creating an interface for it as I think it's somewhat useful.

Signed-off-by: gotjosh <josue@grafana.com>

### CODE REVIEW

File: cluster/cluster.go [hunk 1]
```diff
@@ -33,6 +33,26 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+// ClusterPeer helps consumers gather information about the peer(s) in the cluster
+type ClusterPeer interface {
```
Comment (gotjosh) [inline cluster/cluster.go:37]: I had to place this in the cluster package as both API `v1` and `v2` consume this interface. Once we remove the v1 code I'll move to where it's used in the API package.

Comment (gotjosh) [reply cluster/cluster.go:37]: Also, I'm happy to take suggestions on a better name.

File: cluster/cluster.go [hunk 2]
```diff
@@ -33,6 +33,26 @@ import (
...
+	// Status returns a status string representing the peer state.
+	Status() string
+	// Peers returns the peer nodes in the cluster.
+	Peers() []Node
+}
+
+// Node interface that represents node peers in a cluster
+type Node interface {
+	// String returns the name of the node
+	String() string
```
Comment (pracucci) [inline cluster/cluster.go:51]: I would call it `Name()` for clarity. You would have to wrap `*memberlist.Node` to do it.

File: cluster/cluster.go [hunk 3]
```diff
@@ -33,6 +33,26 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+// ClusterPeer helps consumers gather information about the peer(s) in the cluster
```
Comment (pracucci) [inline cluster/cluster.go:36]:
```suggestion
// Peer is a single peer in a gossip cluster.
```

File: cluster/cluster.go [hunk 4]
```diff
@@ -33,6 +33,26 @@ import (
...
+	// Address returns the IP address of this peer in the cluster.
+	Address() string
+	// Status returns a status string representing the peer state.
+	Status() string
+	// Peers returns the peer nodes in the cluster.
+	Peers() []Node
+}
+
+// Node interface that represents node peers in a cluster
+type Node interface {
```
Comment (pracucci) [inline cluster/cluster.go:49]: Maybe `ClusterMember` could be a better naming, given it represent a member of cluster?

File: cluster/cluster.go [hunk 5]
```diff
@@ -545,6 +565,10 @@ func (p *Peer) Name() string {
 	return p.mlist.LocalNode().Name
 }
 
+func (p *Peer) Address() string {
```
Comment (pracucci) [inline cluster/cluster.go:568]: Who needs / uses this?

File: cluster/cluster.go (hunk 4 above)
Comment (gotjosh) [reply cluster/cluster.go:49]: I like this, thanks!

File: cluster/cluster.go (hunk 5 above)
Comment (gotjosh) [reply cluster/cluster.go:568]: This is used by the API to expose the node IP, but also internally to manage the peers as they join/leave/reconnect.

I have added a comment to signal what this returns.

Comment (gotjosh) [reply cluster/cluster.go:568]: Now I realise what you meant, the actual address is always used from the `Peers()` which returns the `Node` it is not needed in the base peer. Great catch!

**Review by pracucci — APPROVED** [review]
LGTM

File: cluster/cluster.go [hunk 6]
```diff
@@ -591,14 +614,31 @@ func (p *Peer) Self() *memberlist.Node {
 	return p.mlist.LocalNode()
 }
 
+// Member represents a member in the cluster.
+type Member struct {
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply cluster/cluster.go:618]: The linter complains about Exported functions that return unexported types. I thought this was not acceptable to a certain extend?

[the remark this answers is not in the document]
Comment (pracucci) [reply cluster/cluster.go:618]: You can just give a name to `*memberlist.Node`, like:

`node *memberlist.Node`

[the remark this answers is not in the document]
Comment (gotjosh) [reply cluster/cluster.go:618]: oh, you meant specifically the inner attribute - that makes sense.

File: cluster/cluster.go [hunk 7]
```diff
@@ -33,6 +33,29 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 )
 
+// Peer is a single peer in a gossip cluster.
```
Comment (gouthamve) [inline cluster/cluster.go:36]: `Peer` --> `ClusterPeer`

File: cluster/cluster.go [hunk 8]
```diff
@@ -591,14 +614,31 @@ func (p *Peer) Self() *memberlist.Node {
...
+			node: member,
+		})
+	}
+	return peers
 }
 
 // Position returns the position of the peer in the cluster.
 func (p *Peer) Position() int {
-	all := p.Peers()
+	all := p.mlist.Members()
```
Comment (gouthamve) [inline cluster/cluster.go:641]: Why are we not using `p.Peers()` here?

File: cluster/cluster.go [hunk 9]
```diff
@@ -512,15 +535,15 @@ func (p *Peer) peerUpdate(n *memberlist.Node) {
...
 // broadcast messages for the state can be sent.
-func (p *Peer) AddState(key string, s State, reg prometheus.Registerer) *Channel {
+func (p *Peer) AddState(key string, s State, reg prometheus.Registerer) ClusterChannel {
 	p.states[key] = s
 	send := func(b []byte) {
 		p.delegate.bcast.QueueBroadcast(simpleBroadcast(b))
 	}
 	peers := func() []*memberlist.Node {
-		nodes := p.Peers()
+		nodes := p.mlist.Members()
```
Comment (gouthamve) [inline cluster/cluster.go:544]: Same here, when to use `Peers` and when to use `mlist.Members`?

**Review by gouthamve — APPROVED** [review]
LGTM with nits!

File: cluster/cluster.go (hunk 9 above)
Comment (gotjosh) [reply cluster/cluster.go:544]: Does the above make sense?

Comment (gouthamve) [reply cluster/cluster.go:544]: YUSSS

File: cluster/cluster.go (hunk 7 above)
Comment (gotjosh) [reply cluster/cluster.go]: Is this better?

**Pull request merged**